File: thinkbat.py
#!/usr/bin/env python3

import re
import click


class ThinkBat:

    def __init__(self, bat_num=0):
        self._path = "/sys/class/power_supply/BAT" + str(bat_num)
        self._stats = {}
        self.read_stats()

    def _thresh_type_str(self, is_start):
        return "start" if is_start else "end"

    def _thresh_key(self, is_start):
        return self._thresh_type_str(is_start) + "_threshold"

    def _get_thresh_path(self, is_start):
        return self._path + "/charge_control_" + self._thresh_type_str(is_start) + "_threshold"

    def _format_line(self, line):
        return re.sub("\ |\n", "", line).lower()

    def read_stats(self):
        with open(self._path + "/uevent", "r") as f:
            for line in f.readlines():
                line = self._format_line(line)
                key, val = tuple(line.split("="))
                if val.isnumeric():
                    val = int(val)
                key = key.replace("power_supply_", "")
                self._stats[key] = val

        self._stats["capacity"] = 100 * self._stats["energy_full"] / self._stats["energy_full_design"]

        self._stats["percentage"] = 100 * self._stats["energy_now"] / self._stats["energy_full"]

        with open(self._path + "/voltage_now", "r") as f:
            self._stats["voltage"] = int(self._format_line(f.readline())) / 1000000

        def read_thresholds(is_start):
            with open(self._get_thresh_path(is_start), "r") as f:
                self._stats[self._thresh_key(is_start)] = int(self._format_line(f.readline()))

        read_thresholds(is_start=True)
        read_thresholds(is_start=False)

    def set_charge_thresholds(self, start, end):
        def write_threshold(value, is_start):
            try:
                with open(self._get_thresh_path(is_start), "w") as f:
                    f.write(str(value))
                    self._stats[self._thresh_key(is_start)] = value
                    return True
            except PermissionError:
                print("Changing thresholds requires sudo!")
            except Exception as e:
                print("Failed to change thresholds: " + str(e))
            return False

        if write_threshold(start, True) and write_threshold(end, False):
            print("Successfully changed charging thresholds.")

    def man_info(self):
        return "BAT Type: \n" + \
               "\tManufacturer: " + self._stats["manufacturer"] + "\n" + \
               "\tModel: " + self._stats["model_name"] + "\n" + \
               "\tSerial: " + str(self._stats["serial_number"]) + "\n" + \
               "\tTechnology: " + self._stats["technology"]

    def status_info(self):
        return "BAT Status:\n" + \
               "\tPercentage: " + str(round(self._stats["percentage"], 3)) + "%\n" \
               "\tState: -" + self._stats["status"] + "-\n" + \
               "\tCapacity: " + str(round(self._stats["capacity"], 3)) + "%\n" + \
               "\tCycle Count: #" + str(self._stats["cycle_count"]) + "\n" + \
               "\tVoltage: " + str(self._stats["voltage"]) + " V\n" + \
               "\tStart Thresh: " + str(self._stats["start_threshold"]) + "%\n" + \
               "\tStop Thresh: " + str(self._stats["end_threshold"]) + "%"


@click.command()
@click.option("-t", "--thresholds", nargs=2, type=(click.INT, click.INT), help="Set start and stop thresholds")
def prompt(thresholds):
    bat = ThinkBat()
    if thresholds:
        bat.set_charge_thresholds(thresholds[0], thresholds[1])
    else:
        print(bat.man_info())
        print(bat.status_info())


if __name__ == "__main__":
    prompt()

File: thinkbat.c
// Nikolai Herrmann 2024

#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <string.h>


const char *START_THRESH = "/sys/class/power_supply/BAT0/charge_control_start_threshold";
const char *STOP_THRESH = "/sys/class/power_supply/BAT0/charge_control_end_threshold";
const char *STAT_SUMMARY = "/sys/class/power_supply/BAT0/uevent";


int write_num_to_file(int num, const char *file_path)
{
    int file_access = access(file_path, W_OK);
    
    if (file_access != 0)
    {
        if (errno == EACCES)
            printf("Permission denied for: %s\n", file_path);
        else
            printf("File access error for: %s\n", file_path);
        return -1;
    }
    
    FILE *file = fopen(file_path, "w");
    
    if (file == NULL)
    {
        printf("Error opening: %s\n", file_path);
        return -1;
    }

    fprintf(file, "%d", num);
    fclose(file);
    
    return 0;
}

int change_thresholds(int start, int stop)
{
    if (start < stop && (start >= 1 && start <= 100) && (stop >= 1 && stop <= 100))
    {
        if (write_num_to_file(start, START_THRESH) == 0)
            return write_num_to_file(stop, STOP_THRESH);
    }
    else
        printf("Start (%d) and/or stop (%d) threshold not within bounds\n", start, stop);

    return -1;
}

char *split(char *str)
{
    while (*str != '\0')
    {
        if (*str == '=')
            break;
        str++;
    }
    return str + 1;
}

void display_summary(char **metrics, int metric_size)
{
    printf("BAT Type:\n");
    printf("\tManufacturer: %s\n", split(metrics[16]));
    printf("\tModel: %s\n", split(metrics[15]));
    printf("\tSerial: %s\n", split(metrics[17]));

    printf("BAT Status:\n");
    float energy_now = atof(split(metrics[12]));
    float energy_full = atof(split(metrics[11]));
    printf("\tPercentage: %.2f\%\n", (energy_now / energy_full) * 100);
    printf("\tState: %s\n", split(metrics[3]));

    float energy_full_design = atof(split(metrics[10]));
    printf("\tCapacity: %.2f\%\n", (energy_full / energy_full_design) * 100);
    printf("\tCycle count: #%s\n", split(metrics[6]));
    float voltage = atof(split(metrics[8])) / 1000000;
    printf("\tVoltage: %.3f V\n", voltage);
}

int read_digit(const char *file_path, int *digit)
{
    FILE *file = fopen(file_path, "r");

    if (file == NULL)
    {
        printf("Error opening: %s", file_path);
        return -1;
    }

    if (fscanf(file, "%d", digit) == EOF)
    {
        printf("Failed to read digit");
        return -1;
    }

    fclose(file);

    return 0;
}

int display_thresh()
{
    int start;
    int stop;
    if (read_digit(START_THRESH, &start) == 0 && read_digit(STOP_THRESH, &stop) == 0)
    {
        printf("\tStart thresh: %d\%\n", start);
        printf("\tStop thresh: %d\%\n", stop);
        return 0;
    }
    return -1;    
}

void *check_resize(int size, int *max_size, int type_size, void *arr)
{
    if (size > *max_size)
    {
        *max_size *= 2;
        arr = realloc(arr, type_size * *max_size);
    }
    return arr;
}

int stat_summary(const char *file_path)
{
    FILE *file = fopen(file_path, "r");
    
    if (file == NULL)
    {
        printf("Error opening: %s", file_path);
        return -1;
    }
    
    int metric_mem_size = 5;
    int metric_idx = 0;
    char **metrics = malloc(metric_mem_size * sizeof(char *));

    int str_mem_size = 5;
    int str_idx = 0;
    char *str = malloc(str_mem_size * sizeof(char));

    do
    {
        char ch = fgetc(file);
        if (ch == EOF)
            break;
        if (ch == '\n')
        {
            str[str_idx] = '\0';
            metrics = check_resize(metric_idx + 1, &metric_mem_size, sizeof(char *), metrics);
            metrics[metric_idx++] = str;
            str = malloc(str_mem_size * sizeof(char));
            str_idx = 0;
        }
        else 
        {
            str = check_resize(str_idx + 2, &str_mem_size, sizeof(char), str);
            str[str_idx++] = ch;
        }
    } 
    while (1);

    display_summary(metrics, metric_idx);
    display_thresh();

    free(str);
    for (int i = 0; i < metric_idx; i++)
        free(metrics[i]);
    free(metrics);
    fclose(file);

    return 0;
}

int main(int argc, char *argv[])
{
    if (argc >= 3)
    {
        int start = atoi(argv[1]);
        int stop = atoi(argv[2]);
        int change_status = change_thresholds(start, stop);
        printf((change_status == 0) ? "Successfully changed thresholds\n" : "Failed to change thresholds\n");
    }
    else if (argc >= 2 && (strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--help") == 0))
        printf("Usage:  thinkbat [start charging percentage] [stop charging percentage]\n\tthinkbat 75 80\n");
    else
        stat_summary(STAT_SUMMARY);

    return 0;
}